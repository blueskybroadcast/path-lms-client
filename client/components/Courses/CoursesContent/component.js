import React from 'react';
import PropTypes from 'prop-types';

import CoursesPresentationItem from '../CoursesPresentationItem';

const CoursesContent = ({
  coursesIds, coursesData, slug, currency
}) => (
  <div
    className="courses-container"
    data-role="courses-container"
  >
    <div className="courses courses-list ui-sortable">
      {coursesIds && coursesIds.map(id => (
        <CoursesPresentationItem
          key={id}
          slug={slug}
          currency={currency}
          {...coursesData[id]}
        />
      ))}
    </div>
  </div>
);

CoursesContent.propTypes = {
  coursesIds: PropTypes.arrayOf(PropTypes.string),
  coursesData: PropTypes.objectOf(PropTypes.object),
  slug: PropTypes.string.isRequired,
  currency: PropTypes.string.isRequired
};

CoursesContent.defaultProps = {
  coursesIds: null,
  coursesData: null
};

export default CoursesContent;
