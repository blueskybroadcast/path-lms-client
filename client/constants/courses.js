export const COURSE_PRICE_PATTERN = new RegExp('\d+(\.\d{0,2})?');
