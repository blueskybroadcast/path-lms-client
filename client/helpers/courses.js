export const convertGroupsToSellableItems = (groupsIds, groupsData) => {
  const sellableItems = {};
  groupsIds.forEach((id, index) => {
    const group = groupsData[id];
    sellableItems[index] = {
      sellable: true,
      groupId: group.id,
      groupName: group.name,
      labelEnabled: false,
      pricesAttributes: {
        0: {
          currency: 'usd',
          price: '',
          label: ''
        }
      }
    };
  });
  return sellableItems;
};

export const updateCoursesWithIcons = (courses, icons) => {
  const coursesWithIcons = { ...courses };

  icons.forEach((icon) => {
    const parsedContent = icon[0]
      .replace('", ', ' ')
      .replace(', "', ' ')
      .replace('"]', '')
      .split(' ');
    // need to replace it with regex
    const id = parsedContent[1];
    const name = parsedContent[2].toLowerCase();
    const count = icon[1];

    if (coursesWithIcons[id].permittedContent) {
      coursesWithIcons[id].permittedContent[name] = count;
    } else {
      coursesWithIcons[id].permittedContent = {
        [name]: count
      };
    }
  });

  return coursesWithIcons;
};
