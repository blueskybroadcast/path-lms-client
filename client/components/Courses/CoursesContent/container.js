import { connect } from 'react-redux';

import CoursesContent from './component';

import {
  currentAccountSlugSelector, currentAccountCurrencySelector
} from '../../../selectors/authSelectors';
import { coursesIdsSelector, coursesDataSelector } from '../../../selectors/coursesSelectors';

const mapStateToProps = state => ({
  coursesIds: coursesIdsSelector(state),
  coursesData: coursesDataSelector(state),
  slug: currentAccountSlugSelector(state),
  currency: currentAccountCurrencySelector(state)
});

export default connect(mapStateToProps)(CoursesContent);
