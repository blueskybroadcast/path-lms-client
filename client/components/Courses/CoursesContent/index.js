import CoursesContent from './component';

export default CoursesContent;
