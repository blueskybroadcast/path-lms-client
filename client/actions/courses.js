import normalize from 'jsonapi-normalizer';

import {
  ADD_COURSE_REQUEST, ADD_COURSE_SUCCESS,
  FETCH_COURSES_REQUEST, FETCH_COURSES_SUCCESS,
  SORT_COURSES_IN_UI_REQUEST, SORT_COURSES_IN_UI_SUCCESS
} from './types';

import { convertBoolToNumber, formatDate } from '../helpers';
import { updateCoursesWithIcons } from '../helpers/courses';

export const fetchCoursesUserInfo = ids => async (dispatch, getState, api) => {
  const formattedParams = ids.map(id => ({
    id: [id],
    type: 'Course'
  }));
  const res = await api.post('/user_info/load', { icons: formattedParams });
  return res.data.icons;
};

export const fetchCourses = params => async (dispatch, getState, api) => {
  dispatch({ type: FETCH_COURSES_REQUEST });

  await api.get('/courses', { params })
    .then(({ data }) => {
      const courses = normalize(data);
      const { meta } = data;

      const coursesWithIcons = updateCoursesWithIcons(
        courses.entities.course,
        Object.entries(meta.permittedContentCount)
      );

      courses.entities.category = coursesWithIcons;

      if (courses.result.course) {
        dispatch(fetchCoursesUserInfo(courses.result.course)).then((icons) => {
          icons.forEach((icon) => {
            courses.entities.category[icon.id].statuses = {
              purchase: icon.statuses.purchase
            };
          });
          dispatch({
            type: FETCH_COURSES_SUCCESS,
            payload: {
              courses,
              accountBanner: meta.accountBanner,
              coursesDescriptionText: meta.coursesDescriptionText
            }
          });
        });
      } else {
        dispatch({
          type: FETCH_COURSES_SUCCESS,
          payload: {
            courses,
            accountBanner: meta.accountBanner,
            coursesDescriptionText: meta.coursesDescriptionText
          }
        });
      }
    });
};

export const addCourse = data => async (dispatch, getState, api) => {
  dispatch({ type: ADD_COURSE_REQUEST });

  let sellableItems;
  if (data.selectedSellableItems) {
    sellableItems = {
      example_index: {
        sellable: '0',
        group_id: '',
        group_name: '',
        prices_attributes: {
          0: {
            currency: 'usd'
          }
        }
      }
    };
    Object.values(data.selectedSellableItems).forEach((item, index) => {
      sellableItems[index] = {
        sellable: convertBoolToNumber(item.sellable),
        group_id: item.groupId,
        group_name: item.groupName,
        price_attributes: item.pricesAttributes
      };
    });
  }

  await api.post('/courses', {
    course: {
      active: convertBoolToNumber(data.active),
      name: data.name,
      description: data.descriprion,
      cover_photo_url: data.coverPhotoUrl,
      cover_description: data.coverDescription,
      category_ids: data.categoryIds,
      tracks_attributes: data.trackAttributes,
      start_date: formatDate(data.startDate) || undefined,
      end_date: formatDate(data.endDate) || undefined,
      featured: convertBoolToNumber(data.featured),
      fullsize_cover_photo_url: data.fullSizeCoverPhotoUrl,
      show_progress: convertBoolToNumber(data.showProgress),
      search_keywords: data.searchKeywords
    },
    item_admins: {
      admin_ids: data.adminIds
    },
    sellable_items: sellableItems,
    free_cle_settings: {
      use_credits: convertBoolToNumber(data.freeCleEnabled),
      credits_amount: data.freeCleAmount
    },
    sellable_item: {
      disable_purchase: convertBoolToNumber(data.limitPurchaseAvailability),
      disable_purchase_after: data.purchaseAvailabilityDate
        ? formatDate(data.purchaseAvailabilityDate)
        : ''
    },
    expiration: {
      expirable: convertBoolToNumber(data.expirable),
      repurchasable: convertBoolToNumber(data.repurchasable),
      expires_at_strategy: data.expirationByDate ? '1' : '',
      expires_at: formatDate(data.expirationDate),
      days_before_expire_strategy: data.expirationByDays ? '1' : '',
      days_before_expire: data.expirationDays
    }
  }).then(() => {
    dispatch({ type: ADD_COURSE_SUCCESS });
    dispatch(fetchCourses());
  });
};

export const sortCourses = ids => async (dispatch, getState, api) => {
  dispatch({
    type: SORT_COURSES_IN_UI_REQUEST,
    payload: ids
  });
  await api.post('/courses/sort', {
    order: ids
  }).then(() => {
    dispatch({ type: SORT_COURSES_IN_UI_SUCCESS });
  });
};

export const editCourseDescription = text => async (dispatch, getState, api) => {
  await api.post('/account/courses_description', {
    courses_description: {
      text
    }
  }).then(() => dispatch(fetchCourses()));
};
