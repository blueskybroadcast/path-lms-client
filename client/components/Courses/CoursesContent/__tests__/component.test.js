import React from 'react';
import { shallow } from 'enzyme';
import CoursesContent from '../component';

import { coursesIds, coursesData } from '../../../../fixtures/courses';

describe('Courses Content', () => {
  const props = {
    location: { search: 'asd' },
    slug: 'test',
    currency: 'usd'
  };

  test('should render CoursesContent in loading state correctly', () => {
    const wrapper = shallow(
      <CoursesContent
        {...props}
        loading
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render CoursesContent with data correctly', () => {
    const wrapper = shallow(
      <CoursesContent
        {...props}
        loading={false}
        coursesIds={coursesIds}
        coursesData={coursesData}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render CoursesContent without data correctly', () => {
    const wrapper = shallow(
      <CoursesContent
        {...props}
        loading={false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
