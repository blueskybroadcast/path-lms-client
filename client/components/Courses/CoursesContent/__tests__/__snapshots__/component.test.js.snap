// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Courses Content should render CoursesContent in loading state correctly 1`] = `
<div
  className="courses-container"
  data-role="courses-container"
>
  <div
    className="courses-container bsb-loader-container"
    data-role="courses-container"
  >
    <div
      className="bsb-loader"
    >
      <div
        className="loader-wrapper"
      >
        <div
          className="spinner"
        >
          <i
            className="icon icon-spinner icon-spin icon-large"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Courses Content should render CoursesContent with data correctly 1`] = `
<div
  className="courses-container"
  data-role="courses-container"
>
  <div
    className="search-info"
  >
    <div
      className="search-results"
    >
      Search results for '
      '
    </div>
  </div>
  <div
    className="courses courses-list ui-sortable"
  >
    <Component
      backend={[Function]}
    >
      <DropTarget(DragSource(CoursesPresentationItem))
        active={true}
        canDrag={false}
        comments={false}
        courseItemComments={false}
        coverDescription="Test Course 1 cover desctiption"
        coverPhotoUrl="https://cdn.filestackcontent.com/9HnjTksSpyenARCxXMxg"
        currency="usd"
        description={null}
        endDate="2019-03-27"
        expiration={
          Object {
            "id": "3",
            "type": "expiration",
          }
        }
        featured={true}
        fullsizeCoverPhotoUrl="https://cdn.filestackcontent.com/wSqPHkBmRryI2YgTZLxF"
        id="14"
        index={0}
        isAdmin={false}
        key="14"
        moveCard={[Function]}
        name="Test Course 1"
        position={0}
        resizedCoverPhotoUrl="https://cdn.filestackcontent.com/9HnjTksSpyenARCxXMxg/convert?align=left&cache=true&fit=crop&format=jpeg&h=152&quality=100&w=320"
        searchKeywords="quiz, first"
        sellableItems={
          Array [
            Object {
              "id": "10",
              "type": "sellableItem",
            },
          ]
        }
        showProgress={true}
        slug="test"
        startDate="2019-03-13"
        statuses={
          Object {
            "purchase": "unpaid",
          }
        }
      />
      <DropTarget(DragSource(CoursesPresentationItem))
        active={true}
        canDrag={false}
        comments={false}
        courseItemComments={false}
        coverDescription="Test Course 1 cover desctiption"
        coverPhotoUrl="https://cdn.filestackcontent.com/9HnjTksSpyenARCxXMxg"
        currency="usd"
        description={null}
        endDate="2019-04-03"
        expiration={
          Object {
            "id": "4",
            "type": "expiration",
          }
        }
        featured={true}
        fullsizeCoverPhotoUrl="https://cdn.filestackcontent.com/wSqPHkBmRryI2YgTZLxF"
        id="15"
        index={1}
        isAdmin={false}
        key="15"
        moveCard={[Function]}
        name="Test Course 2"
        position={1}
        resizedCoverPhotoUrl="https://cdn.filestackcontent.com/9HnjTksSpyenARCxXMxg/convert?align=left&cache=true&fit=crop&format=jpeg&h=152&quality=100&w=320"
        searchKeywords="quiz, first"
        sellableItems={
          Array [
            Object {
              "id": "11",
              "type": "sellableItem",
            },
          ]
        }
        showProgress={true}
        slug="test"
        startDate="2019-03-24"
        statuses={
          Object {
            "purchase": "unpaid",
          }
        }
      />
    </Component>
  </div>
</div>
`;

exports[`Courses Content should render CoursesContent without data correctly 1`] = `
<div
  className="courses-container"
  data-role="courses-container"
>
  <div
    className="search-info"
  >
    <div
      className="search-results"
    >
      Search results for '
      '
    </div>
  </div>
  <div
    className="courses courses-list ui-sortable"
  >
    <Component
      backend={[Function]}
    />
  </div>
</div>
`;
